src/appleseed/foundation/array/array.h

```diff
@@ -0,0 +1,254 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#pragma once
+
+// appleseed.foundation headers.
+#include "foundation/array/arraytraits.h"
+#include "foundation/platform/compiler.h"
+#include "foundation/utility/alignedallocator.h"
+#include "foundation/utility/test.h"
+
+// appleseed.main headers.
+#include "main/dllsymbol.h"
+
+// Standard headers.
+#include <algorithm>
+#include <cstddef>
+#include <vector>
+
+DECLARE_TEST_CASE(Foundation_Array_Array, MoveConstruct);
+
+namespace foundation
+{
+
+//
+// An array that can hold items of a predetermined set of types.
+//
+
+class APPLESEED_DLLSYMBOL Array
+{
+  public:
+    // Constructor.
+    explicit Array(const ArrayType type, const size_t n = 0);
+
+    // Destructor.
+    ~Array();
+
+    // Copy constructor
+    Array(const Array& other);
+
+    // Move constructor.
+    Array(Array&& other) APPLESEED_NOEXCEPT;
+
+    // Assignment.
+    Array& operator=(const Array& other);
+
+    // Move assignment.
+    Array& operator=(Array&& other) APPLESEED_NOEXCEPT;
+
+    // Return the type of the array elements.
+    ArrayType type() const;
+
+    // Return true if the array is empty.
+    bool empty() const;
+
+    // Return the number of items in the array.
+    size_t size() const;
+
+    // Return the capacity of the array.
+    size_t capacity() const;
+
+    // Remove all items from the array.
+    void clear();
+
+    // Reserve memory for n items.
+    void reserve(const size_t n);
+
+    // Resize the array to n items.
+    void resize(const size_t n);
+
+    // Remove excess capacity from the array.
+    void shrink_to_fit();
+
+    // Equality and inequality tests.
+    bool operator==(const Array& rhs) const;
+    bool operator!=(const Array& rhs) const;
+
+  private:
+    template <typename T>
+    friend class ArrayRef;
+
+    template <typename T>
+    friend class ArrayView;
+
+    GRANT_ACCESS_TO_TEST_CASE(Foundation_Array_Array, MoveConstruct);
+
+    bool is_moved() const;
+
+    const void* begin() const;
+    const void* end() const;
+
+    void* begin();
+    void* end();
+
+    void push_back(const void* p);
+
+    struct Concept
```

Comment:
What's the idea behind the `Concept` and `Model` classes? Why do we need them? Why do we need two?

Reply:
Concept is the abstract base class for all the arrray models. We need it to be able to store different arrays models in the Array and to forward the Array methods to the models. (Concept* m_self)